tests/Feature/Api/CityControllerTest.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Feature\Api;
+
+use App\Models\City;
+use App\Models\Country;
+use App\Models\User;
+use Generator;
+use Laravel\Sanctum\Sanctum;
+use Tests\TestCase;
+
+class CityControllerTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $adminUser = User::factory()->create();
+        Sanctum::actingAs($adminUser, ["HasAdminRole"]);
+    }
+
+    public function testCitiesArrayIsReturned(): void
+    {
+        City::factory()->count(3)->create();
+
+        $response = $this->getJson("/api/admin/cities");
+
+        $response->assertStatus(200)->assertJsonStructure([
+            "cities" => ["*" => []],
+            "providers" => ["*" => []],
+            "countries" => ["*" => []],
+            "citiesWithoutAssignedCountry" => ["*" => []],
+        ]);
+    }
+
+    public function testCityCanBeCreated(): void
+    {
+        $country = Country::factory()->create();
+
+        $city = [
+            "name" => "Legnica",
+            "latitude" => 44.543,
+            "longitude" => -43.122,
+            "country_id" => $country->id,
+        ];
+
+        $this->postJson("/api/admin/cities", $city);
+
+        $this->assertDatabaseHas("cities", $city);
+    }
+
+    public function testCityCannotBeCreatedBecauseFieldsAlreadyExist(): void
+    {
+        $country = Country::factory()->create();
+
+        $city = [
```

Comment:
I don't understand what's going on here

Reply:
It tests if validation works and duplicated city will be created.

Reply:
```
public function testCityCannotBeCreatedBecauseFieldsAlreadyExist(): void
{
    $country = Country::factory()->create();

    $cityData = [
        "name" => "Legnica",
        "latitude" => 44.543,
        "longitude" => -43.122,
        "country_id" => $country->id,
    ];

    $response = $this->postJson("/api/admin/cities", $cityData);

    $response->assertStatus(422);

    $this->assertDatabaseMissing("cities", $cityData);
}
```


For me this :point_up: is more readable, but do it on your feet